### ISSUE TITLE
Enhancement: Dynamic Buffer Groups

### ISSUE THREAD

**Issue by muhmud** [akinsho/bufferline.nvim#347]
Would it be possible to define groups dynamically, rather than statically through config?

**Comment by akinsho** [comment]
Hi @muhmud can you detail exactly what  you're thinking e.g. how will this work? Why would this be valuable, what happens to groups between sessions etc.

**Comment by muhmud** [comment]
Sure, here goes:

What usually ends up happening for me is that I start working on something and buffers start opening up all over the place. Usually, there is a logically grouping that they fall into, such as for a particular Jira ticket, however, it's something that is only relevant for the work I'm doing now, so I want to be able to run a command that prompts me for a name and then (optionally) moves all ungrouped buffers to this new group.

I should then also be able to hide this group completely and start opening buffers again. In this way, I could have multiple groups open at the same time, which effectively become like workspaces, and flip between them as I need to. I think would make working on different projects and tasks a lot more organized.

Although the groups are transient, it would be great if there was an option to save the session (vim natively has a feature like this), so that I could restart the next day with all my previous buffers and groups ready to go. If you didn't save the session, you would just start from scratch.

I think the main difference is that the groups are defined within the editor, rather than in the config.

What do you think?

**Comment by akinsho** [comment]
@muhmud thanks for clarifying. I understand the use case better now. I generally like the idea however one bit of complexity is that when specifying a group you use non simple types e.g.
```lua
  {
    highlight = { guisp = '#51AFEF', gui = 'underline' },
    name = 'tests',
    icon = '',
    matcher = function(buf)
      return buf.filename:match '_spec' or buf.filename:match 'test'
    end,
  },
```
We'd essentially need to take form input which is not easy/possible in nvim and the matcher would like couldn't be a function. Also you wouldn't be able to specify the highlights easily. I guess a solution could be taking the input as a comma separated string and only allowing specifying a regex for the matcher e.g.

```lua
input = "name:jira-tickets,icon:,highlight_guisp:#51AFEF,matcher:'*_spec'"
```
This is kind of ugly and confusing to write out. Alternatively we could open a `vim.ui.input` for every subsequent input which would end up looking like 4 different inputs one after the other.

**Comment by muhmud** [comment]
Yeah, I was thinking about this also when looking through the README.

To keep it simple, I was thinking that everything apart from the name would default to something sensible, so the highlight colour would be some random colour you haven't used yet. Likewise, the icon could be the icon of the buffers that occur most often in the group, and would change as buffers are added. If someone wanted more control, they could run individual commands to override the icon/style/colour, etc., after the group is created. Reducing the configurability and providing really good default behaviour might be the answer.

I think anything more than name and matcher inputs (with the matcher sensibly defaulted) would be too long-winded.

**Comment by akinsho** [comment]
I think defaults are interesting although tbh implementing most of what you describe is actually fairly involved e.g. selecting a random colour, since it's really up to the user to provide those, same as with icons e.g. trying to derive the most common one for a group. I think the icon is optional and can see if there's a better solution later re. colours, a user will have to make up their mind or leave the field blank. I'd generally avoid additional commands etc. as those just add complexity

**Comment by muhmud** [comment]
Do you really think the random colours would be hard? I was thinking have an array of 20 nice looking colours and use `math.random()` to get an index into it, or allocate colours round-robin out of the array as groups are created.

In any case, if there was a way to specify the defaults in the config that would be great, or better still a function in the config so that the defaults could be generated dynamically, then someone could add these features themselves if they wanted to.

For the icons though, there would have to be some sort of event when a buffer is added/removed from a group, so definitely some lift there; just giving you my best case scenario!

**Comment by muhmud** [comment]
Looks pretty good for a v1 :+1: 

I understand about the randomizing going two steps forward instead of one; it would be good to play with this for a while to see what it's like.

**Comment by akinsho** [comment]
Thanks for having a look 👍🏿 yeah, definitely needs a little tweaking / checking stability.

Re. manually adding and removing buffers, I'd say that feature is complex/out of scope for now since that will be a pain to implement. There is no built-in UI to handle multiselection, and I'm dead set against building a custom UI in this plugin for adhoc use cases. I really don't want this to blow up into something super complex.

I guess the functionality could use `vim.ui.select` which only supports selection one at a time, but not sure if that's too laborious or not. In either case feels like this use case easily goes beyond a v1 since it's already beyond what groups offer anyway and is basically a separate feature request i.e. adding buffers to a group which I think is its own can of worms

**Comment by muhmud** [comment]
I just noticed that when you open a new tab, as opposed to a buffer, you get a tab number indicator on the right, which got me thinking...

If there was a config option so that when you move to a different tab, it only showed the buffers from that tab, I think that would pretty much do what I want. The native vim tab functionality would do everything else I want.

One additional nicety would be to be able to associated an icon for each tab, rather than just the number, that would make it perfect.

This is definitely a different direction than before, but might be more suited to this use case.

**Comment by akinsho** [comment]
@muhmud yeah, I think staying on topic regarding dynamic buffer groups would be best. It seems like at the core what you are trying to do is get workspaces working in nvim. I get this issue through this plugin a lot, but I actually think it's not applicable because regardless of what visually you see, nvim *doesn't support workspaces* natively so for example native commands would take you to wherever regardless of which "workspace" you are in. Also, no other plugins would respect this and perhaps most importantly *buffers are not scoped to tabs*, so there's no such concept as a "buffer open in a tab" the buffer isn't open in the tab, it is just open in neovim and visible currently in the tab.

I really want to try to separate all workspace endeavours from this plugin since that goes well beyond a bufferline with nice eye candy into buffer management + bolting on a new concept to neovim which I have seen plugins for, the majority of which have issues. 

To clarify, the point of buffer groups is just to keep things tidy and operate on things as a group, e.g. I want to see all my tests or hide all my tests, I want to see all my docs or hide all my docs etc. It isn't a form of workspace and isn't intended to become a form of workspace.

I think maybe I jumped the gun on implementation since I had an idea for how it could look, but it seems like the core problem to solve with this issue isn't really what I had in mind anyway, and maybe strays in a completely different direction 🤔

**Comment by muhmud** [comment]
I think my requirements and different & the dynamic groups aren't going to cut it, although I think it would still be a useful addition.

About what you were saying about buffers open in a tab, even though buffers are independent of tabs, I can find out which buffers are open in a tab using `tabpagebuflist()`, which would allow for an option to hide buffers in the bufferline that aren't open in the current tab.

Would you accept a PR for that?

**Comment by akinsho** [comment]
@muhmud you can already implement this notion of hiding buffers not available in a tab using a custom filter and the function you pointed out, this has actually been raised before and there's a closed issue with this solution. I don't think anything needs to be added to the plug-in itself for this feature since it's already doable

**Comment by akinsho** [comment]
For reference, the issue I meant was #154

**Comment by muhmud** [comment]
Thanks for the link, I think I could work with this to do what I want :+1: 

One an unrelated note, I've been thinking recently about writing an extension for bufferline that sorts buffers dynamically in MRU order. What are thoughts on this? Would a PR be welcome?

**Comment by akinsho** [comment]
@muhmud there is an open issue for this in #149, would definitely be open to a PR to implement this, I've looked into it myself actually and the main issue that's a little hard to figure out is how to prevent buffers shifting around constantly, there'd have to be some way of weighting changes so that buffers only shift every once in a while rather than reshuffling the entire bufferline on every buf enter

**Comment by akinsho** [comment]
I've decided to close this out since I'm not sure about this feature, not a hard no but having worked on a draft PR for it and given the discussion here I'm much less convinced than I was before but someone feel free to open a PR and change my mind

**Comment by diego-rapoport** [comment]
Any advance on that?
What I thought was just simple commands such as `:BufferlineCreateGroup <name>` and `:BufferlineSendToGroup <name>`(sends the current buffer). The groups would always be inserted on the left, or at least as default for now.
Is that hard to implement and involve some of other issues already discussed? If so, never mind my incomprehension and I'm already sorry for that.

**Comment by akinsho** [comment]
I've closed out my attempts to add this and am not keen on doing this myself anytime soon. The data collection to create a group is really the question. If a contributor can come up with a fully thought out solution (maybe see my closed PR) I'd be happy to review, but I'm not going to take this on myself anytime soon

**Comment by muhmud** [comment]
Hey @akinsho , I thought you might like to know, I put together a solution for MRU switching for vim, as well as tmux: https://github.com/muhmud/switch.

Seems to work well for me.
